packages/flutter/lib/src/widgets/inherited_theme.dart

```diff
@@ -0,0 +1,146 @@
+/// built in, like the contents of a new route or an overlay, will
+/// be able to depend on inherited widget ancestors of the context
+/// it's built in.
+///
+/// {@tool snippet --template=freeform}
+/// This example demonstrates how `InheritedTheme.captureAll()` can be used
+/// to wrap the contents of a new route with the inherited themes that
+/// are present when the route is built - but are not present when route
+/// is actually shown.
+///
+/// If the same code is run without `InheritedTheme.captureAll(), the
+/// new route's Text widget will inherit the "something must be wrong"
+/// fallback text style, rather than the default text style defined in MyApp.
+///
+/// ```dart imports
+/// import 'package:flutter/material.dart';
+/// ```
+///
+/// ```dart main
+/// void main() {
+///   runApp(MyApp());
+/// }
+/// ```
+///
+/// ```dart
+/// class MyAppBody extends StatelessWidget {
+///   @override
+///   Widget build(BuildContext context) {
+///     return GestureDetector(
+///       onTap: () {
+///         Navigator.of(context).push(
+///           MaterialPageRoute(
+///             builder: (BuildContext _) {
+///               // InheritedTheme.captureAll() saves references to themes that
+///               // are found above the context provided to this widget's build
+///               // method, notably the DefaultTextStyle defined in MyApp. The
+///               // context passed to the MaterialPageRoute's builder is not used,
+///               // because its ancestors are above MyApp's home.
+///               return InheritedTheme.captureAll(context, Container(
+///                 alignment: Alignment.center,
+///                 color: Theme.of(context).colorScheme.surface,
+///                 child: Text('Hello World'),
+///               ));
+///             },
+///           ),
+///         );
+///       },
+///       child: Center(child: Text('Tap Here')),
+///     );
+///   }
+/// }
+///
+/// class MyApp extends StatelessWidget {
+///   @override
+///   Widget build(BuildContext context) {
+///     return MaterialApp(
+///       home: Scaffold(
+///         // Override the DefaultTextStyle defined by the Scaffold.
+///         // Descendant widgets will inherit this big blue text style.
+///         body: DefaultTextStyle(
+///           style: TextStyle(fontSize: 48, color: Colors.blue),
+///           child: MyAppBody(),
+///         ),
+///       ),
+///     );
+///   }
+/// }
+/// ```
+/// {@end-tool}
+abstract class InheritedTheme extends InheritedWidget {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const InheritedTheme({
+    Key key,
+    @required Widget child,
+  }) : super(key: key, child: child);
+
+  /// Return a copy of this inherited theme with the specified [child].
+  ///
+  /// If the identical inherited theme is already visible from [context] then
```

Comment:
I'm not sure I understand what "visible" means. Is this to say that if an inherited theme of the same type has already been found, ignore the InheritedThemes of the same type further up the element tree?

Reply:
It means that if the identical inherited theme would be inherited by child, then there's no need to wrap child with the same theme.

For example if Theme.of(context) is non-null then Theme is "visible" in context. If the Theme widget we stashed away in InheritedThemes.captureAll(context, child) is identical to the one that's visible to the returned widget when it's built (in the context where it will actually appear), then there's no need to wrap the child with the saved Theme.